src/commissioner/Discoverer.cpp

```diff
@@ -0,0 +1,264 @@
+#if CHIP_PROGRESS_LOGGING
+    char addressStr[Transport::PeerAddress::kMaxToStringSize];
+    commissionee.mCommissionableNodeAddress.Value().ToString(addressStr);
+    ChipLogProgress(Controller, "Commissionable node candidate identified at %s", addressStr);
+#endif // CHIP_PROGRESS_LOGGING
+    ++mIpIdx;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DnssdDiscoverer::GetValidRecord()
+{
+    while (mNodeIdx < GetDiscoveredNodes().size() && GetDiscoveredNodes().data()[mNodeIdx].IsValid())
+    {
+        if (mIpIdx < GetDiscoveredNodes().data()[mNodeIdx].numIPs)
+        {
+            return CHIP_NO_ERROR;
+        }
+        ++mNodeIdx;
+        mIpIdx = 0;
+    }
+    if (mNodeIdx >= GetDiscoveredNodes().size())
+    {
+        Shutdown(); // If mNodes is full, no point searching further.
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR DnssdDiscoverer::GetAddress(chip::Optional<Transport::PeerAddress> & address,
+                                       chip::Optional<ReliableMessageProtocolConfig> & mrpConfig)
+{
+    if (mNodeIdx >= GetDiscoveredNodes().size() || mIpIdx >= GetDiscoveredNodes().data()[mNodeIdx].numIPs)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    Dnssd::DiscoveredNodeData & node = GetDiscoveredNodes().data()[mNodeIdx];
+    Inet::InterfaceId interfaceId = node.ipAddress[mIpIdx].IsIPv6LinkLocal() ? node.interfaceId[mIpIdx] : Inet::InterfaceId::Null();
+#if 0
+        // TODO(#13326): connecting via TCP does not work
+        if (node.supportsTcp)
+        {
+            address.SetValue(Transport::PeerAddress::TCP(node.ipAddress[mIpIdx], node.port, interfaceId));
+            mrpConfig.SetValue(node.GetMRPConfig());
+        }
+        else
+#endif
+    {
+        address.SetValue(Transport::PeerAddress::UDP(node.ipAddress[mIpIdx], node.port, interfaceId));
+        mrpConfig.SetValue(node.GetMRPConfig());
+    }
+    return CHIP_NO_ERROR;
+}
+
+#endif // CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+
+Discoverer::Discoverer(Controller::DeviceControllerSystemState & systemState, Platform::SharedPtr<SetupPayload> payload,
+                       Delegate * delegate) :
+    Joinable(&mDelegate),
+    mDelegate(delegate), mPayload(payload)
+#if CONFIG_NETWORK_LAYER_BLE
+    ,
+    mBleDiscoverer(systemState, &mDelegate)
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    ,
+    mDnssdDiscoverer(systemState, &mDelegate)
+#endif
+{}
+
+CHIP_ERROR Discoverer::Init()
+{
+    RetainShutdownToken(Platform::MakeShared<ShutdownToken>(&mDelegate));
+    return InProgress() ? CHIP_NO_ERROR : CHIP_ERROR_NO_MEMORY;
+}
+
+void Discoverer::Shutdown()
+{
+#if CONFIG_NETWORK_LAYER_BLE
+    mBleDiscoverer.Shutdown();
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    mDnssdDiscoverer.Shutdown();
+#endif
+    ReleaseShutdownToken();
+}
+
+void Discoverer::SetDelegate(Delegate * delegate)
+{
+    mDelegate = delegate;
+}
+
+CHIP_ERROR Discoverer::Discover()
```

Comment:
Would it make sense to give the caller the ability to separate which transport they'd like to perform their discovery on in situations where both Ble and DNS-SD are available?

Reply:
This is actually doing exactly this.  The Discoverer provides concurrent BLE and DNS-SD discovery when appropriate and available.

The need to perform BLE discovery, DNS-SD discovery and PASE all concurrently turns out to be pretty complex problem to solve.  This is a facility that `src/commissioner/Discoverer.h` + `src/commisisoner/States.h` provides.

I need doxy comments in this file...

Reply:
Comment headers added.